### PR TITLE
Pollute buffer before calling Read()

### PR DESCRIPTION
**Pull request by siying**
Summary:
Add a paranoid check where in case FileSystem layer doesn't fill the buffer but returns succeed, checksum is unlikely to match even if buffer contains a previous block. The byte modified is not useful anyway, so it isn't expect to change any behavior.

Test Plan: See existing CI to pass.

### CODE REVIEW

File: file/random_access_file_reader.cc [hunk 1]
```diff
@@ -214,6 +222,17 @@ IOStatus RandomAccessFileReader::MultiRead(const IOOptions& opts,
                                            AlignedBuf* aligned_buf) const {
   (void)aligned_buf;  // suppress warning of unused variable in LITE mode
   assert(num_reqs > 0);
+
+  // A paranoid check: modify scratch a little bit, so in case underlying
+  // FileSystem doesn't fill the buffer but return succee and `scratch` returns
```
Comment (mrambacher) [inline file/random_access_file_reader.cc:227]: "success"

File: file/random_access_file_reader.cc [hunk 2]
```diff
@@ -41,6 +41,14 @@ IOStatus RandomAccessFileReader::Read(const IOOptions& opts, uint64_t offset,
   (void)aligned_buf;
 
   TEST_SYNC_POINT_CALLBACK("RandomAccessFileReader::Read", nullptr);
+
+  // A paranoid check: modify scratch a little bit, so in case underlying
+  // FileSystem doesn't fill the buffer but return succee and `scratch` returns
```
Comment (mrambacher) [inline file/random_access_file_reader.cc:46]: "success"

File: HISTORY.md [hunk 3]
```diff
@@ -10,6 +10,9 @@
 ### New Features
 * RemoteCompaction's interface now includes `db_name`, `db_id`, `session_id`, which could help the user uniquely identify compaction job between db instances and sessions.
 
+### Miscellaneous
+* Add a paranoid check where in case FileSystem layer doesn't fill the buffer but returns succeed, checksum is unlikely to match even if buffer contains a previous block. The byte modified is not useful anyway, so it isn't expect to change any behavior.
```
Comment (pdillinger) [inline HISTORY.md:14]: Perhaps

    ... isn't expected to change any behavior when FileSystem is satisfying its contract.

**Review by pdillinger — APPROVED** [review]
👍

I'm surprised we don't hit any uninitialized value error in UBSAN but I won't complain if it doesn't

**Review by mrambacher — COMMENTED** [review]
Shouldn't the same be applied to SequentialFileReader as well?

File: file/random_access_file_reader.cc [hunk 4]
```diff
@@ -41,6 +41,14 @@ IOStatus RandomAccessFileReader::Read(const IOOptions& opts, uint64_t offset,
...
+  // contains a previous block, returned value will not pass checksum.
+  if (n > 0 && scratch != nullptr) {
+    scratch[0]++;
```
Comment (anand1976) [inline file/random_access_file_reader.cc:49]: We don't use scratch for direct IO I think. It doesn't apply to Zippy, so should suffice for now. But longer term, we'd need to poison the aligned buffer for direct IO.

File: file/random_access_file_reader.cc [hunk 5]
```diff
@@ -214,6 +222,17 @@ IOStatus RandomAccessFileReader::MultiRead(const IOOptions& opts,
...
   (void)aligned_buf;  // suppress warning of unused variable in LITE mode
   assert(num_reqs > 0);
+
+  // A paranoid check: modify scratch a little bit, so in case underlying
+  // FileSystem doesn't fill the buffer but return succee and `scratch` returns
+  // contains a previous block, returned value will not pass checksum.
+  for (size_t i = 0; i < num_reqs; i++) {
+    FSReadRequest& r = read_reqs[i];
+    if (r.len > 0 && r.scratch != nullptr) {
+      r.scratch[0]++;
```
Comment (anand1976) [inline file/random_access_file_reader.cc:232]: Same comment as above - direct IO doesn't use this scratch buffer

**Pull request closed without merging**